rfcs/20180905-deprecate-collections.md

```diff
@@ -0,0 +1,612 @@
+  ...
+  a = tf.Variable(0)
+  ...
+
+assert tracker.variables == [a]
+```
+
+We might want to expose VariableTracker as well.
+
+
+### **Local variables**
+
+Local variables are primarily used for metrics. With the move to objects for metrics, one can access the variables created for the metrics via the Metric object. 
+
+
+### **Moving average variables**
+
+Collections are used in two places here. We'll remove these calls and ask callers to explicitly pass in a variable list 
+
+
+
+*   [MovingAverageOptimizer.swapping_saver](https://github.com/tensorflow/tensorflow/blob/r1.10/tensorflow/contrib/opt/python/training/moving_average_optimizer.py#L140)
+*   [ExponentialMovingAverage.variables_to_restore](https://github.com/tensorflow/tensorflow/blob/r1.10/tensorflow/python/training/moving_averages.py#L473). We'll have to expose the variables created by ExponentialMovingAverage as they would have to be passed in here for this method.
+
+
+### **Tables**
+
+
+#### **Keras models**
+
+We'll provide a `track_table` method in the backend (similar to `track_variable`) and we'll make sure that when model.fit is called, tables are initialized similar to the way we initialize variables. Changes we'll have to make here are the following
+
+
+
+*   Add a `track_table` method similar to `track_variable` method in keras backend.py
+*   Change feature column code to call `track_table` when it creates tables
+*   Modify `get_session` to initialize the tracked tables in addition to variables
+
+
+#### **Custom graphs**
+
+Similar to `variable_creator_scope` we'll create a `table_creator_scope` that will allow tracking logic to be introduced each time a table is created. This can then be used to collect up all the tables. 
+
+Sample user code might look like
+
+
+```python
+table_creator = tf.TableCreator()
+
+with table_creator:
+  … <graph creation code>
+  table = tf.HashTable(tf.KeyValueTensorInitializer(keys, values), -1)
+  … 
+
+assert table_creator.tables() == [table]
+```
+
+
+We can add some helper methods on TableCreator to collect up all `initializers, vocabulary_filenames`.
+
+
+#### **Alternative proposal for tables**
+
+Initialize tables on first use. Similar to the [OneShotIterator](https://github.com/tensorflow/tensorflow/blob/r1.10/tensorflow/core/ops/dataset_ops.cc#L684), we'll pass in a function argument that can initialize the table. We can even explore the idea of the function to return a dataset and then initialize the table from that dataset. Having a function would avoid overheads of passing in large tensors that might be required to initialize on each lookup (index_table_from_tensor for instance might run into this problem). The Table API would remain the same since we hide initialization from the user and we wouldn't have to run an initializer explicitly. 
+
+The concern with this solution is that in serving, the first request would take more time than the rest of the calls and that sort of inconsistent performance is not desirable. 
```

Comment:
Hidden table initializers have been a persistent thorn for us in some projects.  Having them be explicitly visible at the construction site would be a nice win for readability.